### ISSUE TITLE
default output plugin

### ISSUE THREAD

**Issue by databus23** [vmware/kube-fluentd-operator#17]
As far I see it, its not possible to define a default output plugin for everything.
What I'm looking for is to define something like this centrally after all the namespace specific configuration.
```
<Match **>
@type elasticsearch
host ...
</Match>
```

In our use case namespace or application owners shouldn't need to to know about elasticsearch endpoints, credentials, connection tuning parameters.
By default everything should be shipped as is to a central elasticsearch (or any other output plugin).
Namespace owners can then optional modify, discard, augment events from their application using `<filter>` and `<match>` rules.

Is my assumption correct that this is not possible atm?
Bear with my I'm new to fluentd and the configuration language is giving me headaches.

So I think what would help is a special annotation for configmaps/secrets in kube-system that signals that this config is to be included after the namespace specific stuff in the main `fluentd.conf` file.

What do you think?

**Comment by jvassev** [comment]
@databus23 

What you can do is configure the "global" match at the `kube-system` namespace. If you configure a `<match **>` at the kube-system then it _really_ matches everything - leaving nothing to the namespaces.
See this for example:
https://github.com/vmware/kube-fluentd-operator#a-note-on-the-kube-system-namespace

**Comment by databus23** [comment]
But I want folks to be able to define filter rules and stuff at a namespace level. I just don't want them to need to know how to actually ship the logs in our elastic search (They don't even know the credentials).

**Comment by jvassev** [comment]
Yours is a valid use case. 

At the moment, the only option you have is to instruct namespace owners to copy & paste a segment that reads configuration from the environment using this syntax: https://docs.fluentd.org/v0.12/articles/faq#how-can-i-use-environment-variables-to-configure-parameters-dynamically?
Then, you only need to set the `env` of the fluent container.  However, it's not possible to populate the env using the Helm chart. I've created #18 to track this. I am imagining a solution like this: https://github.com/kubernetes/charts/blob/master/stable/external-dns/values.yaml#L71

The ideal solution would allow an admin to define pseudo plugins at the kube-system level. After that users can just refer to them. For example:
```xml
kube-system.conf:
<plugin databus23>
  @type es
  username admin
  password ....
</plugin>

some-namespace.conf:
<match **>
  @type databus23
</match>
```
Implementing the above is a bit more involved so it will take some time. At the moment you are limited to env vars if you want to conceal the password (quite a lot of code duplication though).

**Comment by jvassev** [comment]
@databus23 
I've implemented the plugin macros feature from  #19 and I'd be happy to gather your feedback :)

Basically, you need to use this image to get the feature `jvassev/kube-fluentd-operator:v1.5.0-2-g025d688`. You should define the "default" output plugin in kube-system like this:
```
# plugin is a non-standard directive ignored by fluentd
<plugin default>
  @type elasticsearch
  host ...
  password ...
</plugin>
```

After that, other namespaces can refer to it easily:
```
<match **>
  @type default
</match>
```

kube-fluentd-operator is just replacing the `<match>` body with the `<plugin>`'s body.

**Comment by jvassev** [comment]
Plugin macros (#19) takes care of this now.
